src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClients/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,256 @@
+                                    throw e;
+                                });
+                            }
+                        }
+                        else if (completedTask.Status == TaskStatus.Canceled)
+                        {
+                            // Throw expected OperationCanceledException
+                            completedTask.GetAwaiter().GetResult();
+                        }
+                    }
+                    else
+                    {
+                        tasks.Remove(parallelAwait);
+                    }
+                }
+            }
+
+            throw lastException;
+        }
+
+        private bool IsRetryableException(AggregateException ex)
+        {
+            int statusCode = 0;
+
+            IReadOnlyCollection<Exception> innerExceptions = ex.Flatten().InnerExceptions;
+
+            if (innerExceptions.Count > 0 && innerExceptions.All(ex => ex is RequestFailedException))
+            {
+                statusCode = (innerExceptions.Last() as RequestFailedException).Status;
+            }
+
+            switch (statusCode)
+            {
+                case (int)HttpStatusCode.RequestTimeout:       // 408
+                case HttpStatusCodeRequestThrottled:           // 429
+                case (int)HttpStatusCode.InternalServerError:  // 500
+                case (int)HttpStatusCode.BadGateway:           // 502
+                case (int)HttpStatusCode.ServiceUnavailable:   // 503
+                case (int)HttpStatusCode.GatewayTimeout:       // 504
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        private IEnumerable<ConfigurationClientState> GetPrioritizedConfigurationClientList()
+        {
+            var startIndex = -1;
+            var clients = new List<ConfigurationClientState>();
+            var i = 0;
+
+            foreach (ConfigurationClientState client in _clients)
+            {
+                if (DateTimeOffset.UtcNow >= client.BackoffEndTime)
+                {
+                    clients.Add(client);
+                    if (startIndex == -1)
+                    {
+                        startIndex = i;
+                    }
+                }
+                ++i;
+            }
+
+            // All configuration clients are in the failed state, so we try all clients regardless.
+            if (startIndex == -1)
+            {
+                clients.AddRange(_clients);
+            }
+            // We have put the available configuration clients in the list first, and populating the rest of the clients even though they might be in the failed state.
+            else if (clients.Count() != _clients.Count())
+            {
+                i = 0;
+                foreach (ConfigurationClientState client in _clients)
+                {
+                    clients.Add(client);
+
+                    if (++i == startIndex)
```

Comment:
So you assume the clients are always ordered by their BackoffEndTime in _clients? Not sure if that's the case given everything is dynamic. Looks like you are actually trying to track the last known good with the startIndex.

Reply:
It is not assumed that `_clients` will have ordered list by `BackoffEndTime`. They will be in the order in which the users provided.
I'm reordering the clients in a way that puts the 'last known good' clients first, followed by clients we know are currently in backoff.

Reply:
Not sure if I follow. It looks to me the startIndex can be just a member variable and we don't really need to make copies of the clients.

Reply:
With the new implementation, startIndex isn't necessary anymore.